### DISCUSSION TITLE
Actions with functional keys

### DISCUSSION THREAD

**Discussion by robsori** [frangoteam/FUXA#1425] [Ideas]
Hi. Many times in practice we encounter the need to use the functional keys F1, F2,...F12 for certain actions. I believe that such configurable keyboard actions for buttons would be useful. Currently there are only mouse actions.

**Comment by unocelli** [comment]
Hi, you mean to bind the function on the application and not on a defined control, right?

**Reply by robsori** [reply]
Hi, I mean the possibility of triggering an action by pressing a functional key on the physical keyboard. This action should be assigned to a button control. This facility would also be useful on touch panels with physical functional keys that are more resistant over time than the touch screens.
<img width="674" alt="Screenshot 2024-10-16 at 16 48 00" src="https://github.com/user-attachments/assets/837fbeb2-e1d3-4bc5-9876-ce78789ea77f">

**Comment by unocelli** [comment]
Ok, then the page visualised have to listen the pressed Function keys

**Reply by robsori** [reply]
Yes, in addition to mouse actions, events can also be assigned to the pressing of the functional keys on the physical keyboard.
![Screenshot 2024-10-17 at 09 41 47](https://github.com/user-attachments/assets/a0d52502-328f-4d66-932d-150b1713a31b)

**Reply by MatthewReed303** [reply]
@robsori I have been thinking about this one and I thought a good solution would be a keyboard driver, like a PLC driver etc but the keys are actually Tags that can be used anywhere in Fuxa. Would offer a lot more flexibility in what you could do, could use simple Industrial keyboards, num pads etc  The issue here is do we use the client front end to do this, we can't use standard keys, will arrow keys etc cause undesired behavior, or do we use just the node.js server and ability to select the USB input device? For example F keys already do stuff in the browser that would cause undesired behavior etc we need a way to isolate the keyboard from the OS also so it only interacts with the desired application. I had some matrix keyboard I was going to use, use an micro controller STM32 etc to read it and convert keystrokes to Modbus TCP or something, buffer key strokes in the MC and send over Modbus

**Comment by MatthewReed303** [comment]
@unocelli I spoke to you a while ago about something similar binding physical key presses and also using arrow keys to navigate input fields etc and use enter to enter the input field and enter again to confirm and exit or any arrow key will just exit the field without applying anything.

**Reply by unocelli** [reply]
@MatthewReed303 does not seem to me to have had to do with function keys

**Reply by MatthewReed303** [reply]
@unocelli no wasn't specifically function keys, but similar integration would be required to use keys from the keyboard. So would be best to attack both situations with the same integration.

**Reply by unocelli** [reply]
I think function keys are too specific to the hardware being used. FUXA to which one and how should the F[n] event be sent?
